CPS-XXXX/README.md

```diff
@@ -0,0 +1,91 @@
+- Ledger   | For proposals regarding the Cardano ledger (including Reward Sharing Schemes)
+- Catalyst | For proposals affecting Project Catalyst / the Jörmungandr project
+
+-->
+
+## Abstract
+
+Determinism of transactions gives rise to valuable ledger properties, such as:
+
+- atomicity (a transaction is either fully accepted or fully rejected without a fee)
+- predictability of monetary changes and fees
+
+However, it comes at a cost:
+
+- Development becomes more complicated because of the need to "re-invent" non-determinism where it is required for the dApp logic
+- UTxO contention limits the number of concurrent users a dApp can have
+
+Full determinism *may* be unnecessarily strict as a way to achieve the "good" properties, as the design space for the alternatives haven't been fully explored.
+
+## Problem
+
+Let's consider two types of determinism:
+
+I. **execution layer determinism**: can we predict the exact result of a *transaction*, if it passes? Enforced on Cardano, does not exist on EVM.
+
+II. **application layer determinism**: can we predict the outcome of a *user action* within a dApp, if it succeeds?
+
+Whenever app layer (type-II) non-determinism is desired, the go-to solution for Cardano developers is to split one non-deterministic *app action* into multiple *transactions*, typically via on-chain orders and order batchers.
+
+Batchers establish ordering and execute concurrent orders from different users. Orders are typically implemented as transactions that create UTxOs with inline datums. Batchers would return the funds if it's impossible to execute an order, and the user has to compensate the tx fee for the batcher in advance.
+
+This solution introduces a whole set of new problems:
+
+- **Uncertain protocol liveness**. Batchers depend on someone initiating the executing transaction. The node does not provide the guarantee that a valid transaction will pass within a certain timeframe, therefore the risk of order getting stuck has to be accounted for by the user: *no immediate execution is provided*.
+- **Batcher fees and min-ada compensation**. The cost of initiating a transaction has to be compensated to the batcher. The batcher, no matter how efficient, can't take less than min-ada value (the cost of creating a UTxO), because that is the minimum ADA value a UTxO can have. The change can't be returned to the order initiator easily, because it is usually significantly lower than the min-ada.
+- **Doubled action confirmation time**. Two transactions take more time to pass, even assuming immediate order execution.
+- **Incentives for centralization**. Although decentralized batchers are possible, in reality designing them is more complex than building a centralized batcher bot. Decentralized batchers *still* don't provide liveness guarantees: it *may just happen* that no executor bots will be available at any given time, despite the incentives.
+
+## Use cases
+
+### Automated market maker DEX
```

Comment:
> They directly undermine the security assumptions of Proof-of-Stake since users are forced to give up most, if not all, delegation control, voting control, and custody of their assets

It's possible to use own staking credentials with dApp's script as payment credential

> They do not allow users to specify their own prices which makes them extremely economically inefficient

Existing AMM DEXes allow placing orders as a side-feature. An order book existing alongside an AMM is how  [uniswap](https://app.uniswap.org/limit) does that.

> They are fundamentally centralized since updating/maintaining the liquidity pools is controlled by nothing more than a multisig; governance actions are not trustlessly enforced

Any governance scheme can be attached to any protocol.

Liquidity pools without an update mechanism are possible. An order book that lets the admins do something with the orders is also possible.

Reply:
Anyway, AMM DEX is just an example here. An order book suffers from the same problem: an order can be matched only by one counter-party.